lib/msf/ui/console/command_dispatcher/evasion.rb

```diff
@@ -0,0 +1,124 @@
+module Msf
+module Ui
+module Console
+module CommandDispatcher
+class Evasion
+
+  include Msf::Ui::Console::ModuleCommandDispatcher
+
+  def commands
+    super.update({
+      'run'      => 'Launches the evasion module',
+      'rerun'    => 'Reloads and launches the evasion module',
+      'exploit'  => 'This is an alias for the run command',
+      'rexploit' => 'This is an alias for the rerun command',
+      'reload'   => 'Reloads the auxiliary module'
+    }).merge(mod ? mod.evasion_commands : {})
+  end
+
+  def name
+    'Evasion'
+  end
+
+  def cmd_run(*args)
+    opts = {
+      'Encoder'    => mod.datastore['ENCODER'],
+      'Payload'    => mod.datastore['PAYLOAD'],
+      'Nop'        => mod.datastore['NOP'],
+      'LocalInput' => driver.input,
+      'LocalOutput' => driver.output
+    }
+
+    begin
+      mod.run_simple(opts)
+    rescue ::Interrupt
+      print_error('Evasion interrupted by the console user')
+    rescue ::Exception => e
+      print_error("Evasion failed: #{e.class} #{e}")
+      elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}")
+    end
+  end
+
+  alias cmd_exploit cmd_run
+
+  def cmd_rerun(*args)
+    if reload(true)
+      cmd_run(*args)
+    end
+  end
+
+  alias cmd_rexploit cmd_rerun
+
+  def cmd_exploit_tabs(str, words)
+    fmt = {
+      '-e' => [ framework.encoders.map { |refname, mod| refname } ],
+      '-f' => [ nil                                               ],
+      '-h' => [ nil                                               ],
+      '-j' => [ nil                                               ],
+      '-J' => [ nil                                               ],
+      '-n' => [ framework.nops.map { |refname, mod| refname }     ],
+      '-o' => [ true                                              ],
+      '-p' => [ framework.payloads.map { |refname, mod| refname } ],
+      '-t' => [ true                                              ],
+      '-z' => [ nil                                               ]
+    }
+    tab_complete_generic(fmt, str, words)
+  end
+
+  private
+
+  def self.choose_payload(mod, target)
+
+    # Choose either the real target or an invalid address
+    # This is used to determine the LHOST value
+    rhost = mod.datastore['RHOST'] || '50.50.50.50'
+
+    # A list of preferred payloads in the best-first order
+    pref = [
```

Comment:
It's not clear if we need this for an evasion module, since we're not trying to pair it with an exploit.

Reply:
Good catch there. Evasion works best with encrypted payloads and I forgot to actually use this method. Should work now. Thanks.